File: app/src/main/java/com/benny/pxerstudio/activity/ProjectManagerActivity.java
package com.benny.pxerstudio.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.benny.pxerstudio.util.AdHelper;
import com.benny.pxerstudio.R;
import com.benny.pxerstudio.util.Tool;
import com.benny.pxerstudio.pxerexportable.ExportingUtils;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.NativeExpressAdView;
import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.IAdapter;
import com.mikepenz.fastadapter.adapters.ItemAdapter;
import com.mikepenz.fastadapter.items.AbstractItem;

import java.io.File;
import java.io.FileFilter;
import java.util.ArrayList;
import java.util.List;

public class ProjectManagerActivity extends AppCompatActivity {

    ArrayList<File> projects = new ArrayList<>();
    FastAdapter<Item> fa;
    ItemAdapter<Item> ia;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_project_manager);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        RecyclerView rv = (RecyclerView) findViewById(R.id.rv);

        //Comment this line out and the if statement if you forked this repo or downloaded the code
        NativeExpressAdView adView = AdHelper.checkAndEnableAd(this);
        if (adView != null){
            final FrameLayout fl = new FrameLayout(this);
            fl.setVisibility(View.GONE);

            adView.setAdListener(new AdListener(){
                @Override
                public void onAdLoaded() {
                    fl.setVisibility(View.VISIBLE);
                    super.onAdLoaded();
                }
            });

            //ProgressBar progressBar = new ProgressBar(this,null,android.R.attr.progressBarStyle);
            //progressBar.setIndeterminate(true);

            RelativeLayout.LayoutParams lp = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            lp.addRule(RelativeLayout.CENTER_HORIZONTAL);
            lp.addRule(RelativeLayout.ALIGN_PARENT_TOP);

            FrameLayout.LayoutParams lp2 = new FrameLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            lp2.gravity = Gravity.CENTER;

            //fl.addView(progressBar,lp2);
            fl.addView(adView);
            ((RelativeLayout)findViewById(R.id.content_project_manager)).addView(fl,lp);

            fl.setId(R.id.ad_view);
            ((RelativeLayout.LayoutParams)rv.getLayoutParams()).addRule(RelativeLayout.BELOW,R.id.ad_view);
        }

        rv.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        fa = new FastAdapter<>();
        ia = new ItemAdapter<>();

        fa.withSelectable(false);
        rv.setAdapter(ia.wrap(fa));

        projects.clear();

        //Find all projects
        final File parent = new File(ExportingUtils.getProjectPath());
        if (parent.exists()){
            File[] temp = parent.listFiles(new PxerFileFilter());
            for (int i = 0; i < temp.length; i++) {
                projects.add(temp[i]);
            }
            if (projects.size() >= 1) {
                findViewById(R.id.noProjectFound).setVisibility(View.GONE);

                for (int i = 0; i < projects.size(); i++) {
                    String mName = projects.get(i).getName().substring(0, projects.get(i).getName().lastIndexOf('.'));
                    String mPath = projects.get(i).getPath();
                    ia.add(new Item(mName, mPath));
                }

                fa.withOnClickListener(new FastAdapter.OnClickListener<Item>() {
                    @Override
                    public boolean onClick(View v, IAdapter<Item> adapter, Item item, int position) {
                        Intent newIntent = new Intent();
                        newIntent.putExtra("selectedProjectPath",item.path);

                        setResult(RESULT_OK,newIntent);
                        finish();
                        return true;
                    }
                });

                fa.withOnLongClickListener(new FastAdapter.OnLongClickListener<Item>() {
                    @Override
                    public boolean onLongClick(View v, IAdapter<Item> adapter, Item item, final int position) {
                        PopupMenu pm = new PopupMenu(v.getContext(),v);
                        pm.inflate(R.menu.menu_popup_project);
                        pm.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                            @Override
                            public boolean onMenuItemClick(MenuItem item) {
                                switch (item.getItemId()){
                                    case R.id.rename:
                                        Tool.promptTextInput(ProjectManagerActivity.this,getString(R.string.rename)).input(null, projects.get(position).getName(), false, new MaterialDialog.InputCallback() {
                                            @Override
                                            public void onInput(@NonNull MaterialDialog dialog, CharSequence input) {
                                                String mInput = input.toString();
                                                if (!mInput.endsWith(".pxer"))
                                                    mInput += ".pxer";

                                                File fromFile = new File(projects.get(position).getPath());
                                                File newFile = new File(projects.get(position).getParent(),mInput);

                                                if (fromFile.renameTo(newFile)) {
                                                    projects.set(position,newFile);
                                                    ia.set(position, new Item(newFile.getName(),newFile.getPath()));
                                                    fa.notifyAdapterItemChanged(position);

                                                    Intent newIntent = new Intent();
                                                    newIntent.putExtra("fileNameChanged",true);

                                                    setResult(RESULT_OK,newIntent);
                                                }
                                            }
                                        }).show();
                                        break;
                                    case R.id.delete:
                                        Tool.prompt(ProjectManagerActivity.this).title(R.string.deleteproject).content(R.string.deleteprojectwarning).positiveText(R.string.delete).onPositive(new MaterialDialog.SingleButtonCallback() {
                                            @Override
                                            public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                                                if (projects.get(position).delete()){
                                                    ia.remove(position);
                                                    projects.remove(position);

                                                    if (projects.size() < 1)
                                                        findViewById(R.id.noProjectFound).setVisibility(View.VISIBLE);

                                                    Intent newIntent = new Intent();
                                                    newIntent.putExtra("fileNameChanged",true);

                                                    setResult(RESULT_OK,newIntent);

                                                    Tool.toast(ProjectManagerActivity.this,getString(R.string.projectdeleted));
                                                }else
                                                    Tool.toast(ProjectManagerActivity.this,getString(R.string.unabletodeleteproject));
                                            }
                                        }).show();
                                        break;
                                }
                                return true;
                            }
                        });
                        pm.show();
                        return true;
                    }
                });
            }
        }
    }

    public class PxerFileFilter implements FileFilter{
        @Override
        public boolean accept(File pathname) {
            if (pathname.getName().endsWith(".pxer")) return true;
            return false;
        }
    }

    public static class Item extends AbstractItem<ProjectManagerActivity.Item, ProjectManagerActivity.Item.ViewHolder> {

        public String title;
        public String path;

        public Item(String title,String path){
            this.title = title;
            this.path = path;
        }

        @Override
        public int getType() {
            return 0;
        }

        @Override
        public int getLayoutRes() {
            return R.layout.item_projectitem;
        }

        @Override
        public ViewHolder getViewHolder(View v) {
            return new ViewHolder(v);
        }

        @Override
        public void bindView(ViewHolder holder, List payloads) {
            super.bindView(holder, payloads);

            holder.projectTitle.setText(title);
            holder.projectPath.setText(path);
        }

        public static class ViewHolder extends RecyclerView.ViewHolder {
            TextView projectTitle;
            TextView projectPath;

            ViewHolder(View view) {
                super(view);

                projectTitle = (TextView) view.findViewById(R.id.title);
                projectPath = (TextView) view.findViewById(R.id.path);
            }
        }
    }

}

File: app/src/main/java/com/benny/pxerstudio/activity/ProjectManagerActivity.kt
package com.benny.pxerstudio.activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.PopupMenu
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.recyclerview.widget.LinearLayoutManager
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.input
import com.benny.pxerstudio.R
import com.benny.pxerstudio.databinding.ActivityProjectManagerBinding
import com.benny.pxerstudio.databinding.ItemProjectBinding
import com.benny.pxerstudio.pxerexportable.ExportingUtils
import com.benny.pxerstudio.util.Tool
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.adapters.ItemAdapter
import com.mikepenz.fastadapter.items.AbstractItem
import com.mikepenz.fastadapter.select.getSelectExtension
import java.io.File
import java.io.FileFilter
import java.util.*

class ProjectManagerActivity : AppCompatActivity() {

    private var projects = ArrayList<File>()
    private lateinit var fa: FastAdapter<Item>
    private lateinit var ia: ItemAdapter<Item>
    private lateinit var binding: ActivityProjectManagerBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityProjectManagerBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.projectManagerToolbar)

        binding.projectManagerCM.cMRecyclerView.layoutManager =
            LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        ia = ItemAdapter()
        fa = FastAdapter.with(ia)

        fa.getSelectExtension().isSelectable = false
        binding.projectManagerCM.cMRecyclerView.adapter = fa

        projects.clear()

        //Find all projects
        val parent = File(ExportingUtils.getProjectPath(this))
        if (parent.exists()) {
            val temp = parent.listFiles(PxerFileFilter())
            for (i in temp!!.indices) {
                projects.add(temp[i])
            }
            if (projects.size >= 1) {
                binding.projectManagerCM.cMNoProjectFound.isGone = true

                for (i in projects.indices) {
                    val mName = projects[i].name.substring(0, projects[i].name.lastIndexOf('.'))
                    val mPath = projects[i].path
                    ia.add(Item(mName, mPath))
                }

                fa.onClickListener = { _, _, item, _ ->
                    val newIntent = Intent()
                    newIntent.putExtra("selectedProjectPath", item.path)

                    setResult(RESULT_OK, newIntent)
                    finish()
                    true
                }

                fa.onLongClickListener = { v, _, _, position ->
                    val pm = PopupMenu(v.context, v)
                    pm.inflate(R.menu.menu_popup_project)
                    pm.setOnMenuItemClickListener { item ->
                        when (item.itemId) {
                            R.id.menu_popup_project_rename ->
                                MaterialDialog(this).show {
                                    title = getString(R.string.rename)
                                    input(hint = projects[position].name) { _, text ->
                                        var mInput = "$text"
                                        if (!mInput.endsWith(".pxer"))
                                            mInput += ".pxer"

                                        val fromFile = File(projects[position].path)
                                        val newFile = File(projects[position].parent, mInput)

                                        if (fromFile.renameTo(newFile)) {
                                            projects[position] = newFile
                                            ia[position] = Item(newFile.name, newFile.path)
                                            fa.notifyAdapterItemChanged(position)

                                            val newIntent = Intent()
                                            newIntent.putExtra("fileNameChanged", true)

                                            setResult(RESULT_OK, newIntent)
                                        }
                                    }
                                    positiveButton(android.R.string.ok)
                                }

                            R.id.menu_popup_project_delete -> Tool.prompt(this@ProjectManagerActivity)
                                .title(R.string.delete_project)
                                .message(R.string.delete_project_warning)
                                .positiveButton(R.string.delete).positiveButton {
                                    if (projects[position].delete()) {
                                        ia.remove(position)
                                        projects.removeAt(position)

                                        if (projects.size < 1)
                                            binding.projectManagerCM.cMNoProjectFound.isVisible =
                                                true

                                        val newIntent = Intent()
                                        newIntent.putExtra("fileNameChanged", true)

                                        setResult(RESULT_OK, newIntent)

                                        Tool.toast(
                                            this@ProjectManagerActivity,
                                            getString(R.string.project_deleted)
                                        )
                                    } else
                                        Tool.toast(
                                            this@ProjectManagerActivity,
                                            getString(R.string.error_deleting_project)
                                        )
                                }.show()
                        }
                        true
                    }
                    pm.show()
                    true
                }
            }
        }
    }

    inner class PxerFileFilter : FileFilter {
        override fun accept(pathname: File): Boolean {
            return pathname.name.endsWith(".pxer")
        }
    }

    class Item(var title: String, var path: String) : AbstractItem<Item.ViewHolder>() {

        override val type: Int
            get() = 0

        override val layoutRes: Int
            get() = R.layout.item_project

        override fun getViewHolder(v: View): ViewHolder {
            return ViewHolder(v)
        }

        class ViewHolder internal constructor(view: View) : FastAdapter.ViewHolder<Item>(view) {
            private var itemProjectBinding = ItemProjectBinding.bind(view)

            override fun bindView(item: Item, payloads: List<Any>) {
                itemProjectBinding.itemProjectTitle.text = item.title
                itemProjectBinding.itemProjectPath.text = item.path
            }

            override fun unbindView(item: Item) {
            }
        }
    }
}
